ddsrouter_core/src/cpp/writer/implementations/rtps/CommonWriter.cpp

```diff
@@ -0,0 +1,356 @@
+}
+
+// Specific enable/disable do not need to be implemented
+utils::ReturnCode CommonWriter::write_(
+        std::unique_ptr<DataReceived>& data) noexcept
+{
+
+    // Take new Change from history
+    fastrtps::rtps::CacheChange_t* new_change = rtps_writer_->new_change(eprosima::fastrtps::rtps::ChangeKind_t::ALIVE);
+
+    // If still is not able to get a change, return an error code
+    if (!new_change)
+    {
+        return utils::ReturnCode::RETCODE_ERROR;
+    }
+
+    // Get the Payload without copy only if it has length
+    if (data->payload.length > 0)
+    {
+        eprosima::fastrtps::rtps::IPayloadPool* payload_owner = payload_pool_.get();
+        if (!payload_pool_->get_payload(data->payload, payload_owner, (*new_change)))
+        {
+            logDevError(DDSROUTER_RTPS_WRITER, "Error getting Payload.");
+            return utils::ReturnCode::RETCODE_ERROR;
+        }
+    }
+
+
+    logDebug(DDSROUTER_RTPS_WRITER,
+            "CommonWriter " << *this << " sending payload " << new_change->serializedPayload << " from " <<
+            data->qos.source_guid);
+
+    // Fill cache change with specific data to send
+    fill_to_send_data_(new_change, data);
+
+    // Get params to write (if set)
+    eprosima::fastrtps::rtps::WriteParams write_params;
+
+    // RPC support
+    // If writer params has been set specifically, use them in change
+    if (data->qos.write_params.is_set())
+    {
+        write_params.related_sample_identity(data->qos.write_params.value.related_sample_identity());
+    }
+
+    // Send data by adding it to CommonWriter History
+    rtps_history_->add_change(new_change, write_params);
+
+    // At this point, write params is now the output of adding change
+    fill_sent_data_(write_params, data);
+
+    // TODO: check if this being synchronous means that we can remove the change in case of volatile and/or best_effort
+
+    // When max history size is reached, remove oldest cache change
+    if (rtps_history_->isFull())
+    {
+        rtps_history_->remove_min_change();
+    }
+
+    return utils::ReturnCode::RETCODE_OK;
+}
+
+void CommonWriter::fill_to_send_data_(
+        fastrtps::rtps::CacheChange_t* to_send_change_to_fill,
+        std::unique_ptr<types::DataReceived>& data) const noexcept
+{
+    if (repeater_)
+    {
+        // Add origin to change in case the cache change is RouterCacheChange (only in repeater mode)
+        types::RouterCacheChange& change_ref = static_cast<types::RouterCacheChange&>(*to_send_change_to_fill);
+        change_ref.last_writer_guid_prefix = data->qos.source_guid.guidPrefix;
+    }
+
+    // Set keys in case topic has keys
+    if (topic_.keyed)
+    {
+        to_send_change_to_fill->instanceHandle = data->qos.instanceHandle;
+    }
+}
+
+void CommonWriter::fill_sent_data_(
+        const eprosima::fastrtps::rtps::WriteParams& params,
+        std::unique_ptr<types::DataReceived>& data_to_fill) const noexcept
```

Comment:
Perhaps it would be better to pass the cache change, as it contains more information that could be useful in the future, like timestamp.

Reply:
At the time of calling this, the cache change is not available anymore (it is, but probably it has been already copied or owned somewhere else).

Reply:
OK.
With respect to timestamp, I believe we should propagate the source_timestamp in order to support QoS such as lifespan.